### ISSUE TITLE
MONGODB_ROOT_PASSWORD have to be set on a 'primary' node error while using bitbucket pipelines

### ISSUE THREAD

**Issue by B-son** [bitnami/bitnami-docker-mongodb#151]
**Description**
While using bitbucket pipelines and mongodb with replica set, during setup of db error occures
with these additional flags in bitbucket-pipelines.yml
environment:
        MONGODB_REPLICA_SET_MODE: 'primary'
        MONGODB_REPLICA_SET_NAME: 'rs0'

**Describe the results you received:**
In logs mongo logs: 
```
INFO  ==> ** Starting MongoDB setup **
INFO  ==> Validating
INFO  ==> Validating settings in MONGODB_* env vars...
ERROR ==> MONGODB_ROOT_PASSWORD have to be set on a 'primary' node!
```
 which results in failing deployment

**Additional information you deem important (e.g. issue happens only occasionally):**
I've checked latest, 4.1.13-debian-9-r24 and 4.0.10 images, all have the same issue. Problem showed up today, last weeks deployments were fine.
**Version**

- Output of `docker version`:

```
N/A
```

- Output of `docker info`:

```
N/A
```

- Output of `docker-compose version` (if applicable):

```
N/A
```

**Additional environment details (AWS, VirtualBox, Docker for MAC, physical, etc.):**

**Comment by marcosbc** [comment]
Hi @B-son, just to clarify, note that in a primary node you must set the `MONGODB_ROOT_PASSWORD` environment variable and in a secondary/arbiter node, you must set `MONGODB_PRIMARY_ROOT_PASSWORD`.

It seems this error was introduced last week, when we published the Bitnami MongoDB container based on Bash logic (which is open-source and available in this repository). In the previous container release, this parameter was taken from either of those if one of those wasn't empty, which is not the ideal solution.

Could you check your container deployment configuration and add the proper environment variables (even though this is a breaking change between the releases)?

**Comment by B-son** [comment]
Hi, I've checked deployment with those flags and I'm getting:
```
ERROR ==> In order to configure MongoDB replica set authentication you need to provide the MONGODB_REPLICA_SET_KEY on every node, specify MONGODB_ROOT_PASSWORD in the primary node and MONGODB_PRIMARY_ROOT_PASSWORD in the rest of nodes
```
I've also tried to running those flags with setting `MONGODB_PRIMARY_HOST` to localhost:27017 and I've got the same error.

**Comment by marcosbc** [comment]
It seems like you are not configuring replication properly, check [this section](https://github.com/bitnami/bitnami-docker-mongodb#setting-up-a-replication) in the README for more information on how to do so.

There is also a `docker-compose.yml` example which configures all required environment variables for a replicaset [here](https://github.com/bitnami/bitnami-docker-mongodb/blob/master/docker-compose-replicaset.yml), which should be enough to get you started.

If you keep getting problems, could you share with us the environment variables you are configuring? With sample/dummy values if possible.

**Comment by misuzu** [comment]
Hello. Same issue here. Used this config, everything was working for almost a year:
```
  mongodb-primary:
    image: bitnami/mongodb:latest
    network_mode: host
    environment:
      - MONGODB_REPLICA_SET_MODE=primary
      - MONGODB_REPLICA_SET_NAME=test123
```
I have added ```MONGODB_REPLICA_SET_KEY``` and ```MONGODB_ROOT_PASSWORD``` and not able to connect to database anymore.

```
  mongodb-primary:
    image: bitnami/mongodb:latest
    network_mode: host
    environment:
      - MONGODB_REPLICA_SET_KEY=test123
      - MONGODB_REPLICA_SET_MODE=primary
      - MONGODB_REPLICA_SET_NAME=test123
      - MONGODB_ROOT_PASSWORD=password
```

```
% docker run -it --rm --net=host -t bitnami/mongodb:latest mongo mongodb://127.0.0.1/default?replicaSet=test123 

Welcome to the Bitnami mongodb container
Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-mongodb
Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-mongodb/issues
Send us your feedback at containers@bitnami.com


MongoDB shell version v4.0.10
connecting to: mongodb://127.0.0.1:27017/default?gssapiServiceName=mongodb&replicaSet=test123
2019-06-26T07:50:20.464+0000 I NETWORK  [js] Starting new replica set monitor for test123/127.0.0.1:27017
2019-06-26T07:50:20.464+0000 I NETWORK  [js] Successfully connected to 127.0.0.1:27017 (1 connections now open to 127.0.0.1:27017 with a 5 second timeout)
2019-06-26T07:50:20.464+0000 I NETWORK  [js] changing hosts to test123/127.0.1.1:27017 from test123/127.0.0.1:27017
2019-06-26T07:50:20.465+0000 I NETWORK  [js] Successfully connected to 127.0.1.1:27017 (1 connections now open to 127.0.1.1:27017 with a 5 second timeout)
Implicit session: session { "id" : UUID("18c8f531-49be-4640-a5cd-601008601d3d") }
MongoDB server version: 4.0.10
Welcome to the MongoDB shell.
For interactive help, type "help".
For more comprehensive documentation, see
        http://docs.mongodb.org/
Questions? Try the support group
        http://groups.google.com/group/mongodb-user
2019-06-26T07:50:20.466+0000 I STORAGE  [main] In File::open(), ::open for '//.mongorc.js' failed with Unknown error
test123:PRIMARY> db.test.find()
Error: error: {
        "operationTime" : Timestamp(1561535433, 1),
        "ok" : 0,
        "errmsg" : "command find requires authentication",
        "code" : 13,
        "codeName" : "Unauthorized",
        "$clusterTime" : {
                "clusterTime" : Timestamp(1561535433, 1),
                "signature" : {
                        "hash" : BinData(0,"1KjRlvdTJz1sg+Auie+fjYZ0Lp4="),
                        "keyId" : NumberLong("6706743530380853249")
                }
        }
}
test123:PRIMARY>
```

```
% docker run -it --rm --net=host -t bitnami/mongodb:latest mongo mongodb://root:password@127.0.0.1/default?replicaSet=test123

Welcome to the Bitnami mongodb container
Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-mongodb
Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-mongodb/issues
Send us your feedback at containers@bitnami.com


MongoDB shell version v4.0.10
connecting to: mongodb://127.0.0.1:27017/default?gssapiServiceName=mongodb&replicaSet=test123
2019-06-26T07:51:02.497+0000 I NETWORK  [js] Starting new replica set monitor for test123/127.0.0.1:27017
2019-06-26T07:51:02.497+0000 I NETWORK  [js] Successfully connected to 127.0.0.1:27017 (1 connections now open to 127.0.0.1:27017 with a 5 second timeout)
2019-06-26T07:51:02.498+0000 I NETWORK  [js] changing hosts to test123/127.0.1.1:27017 from test123/127.0.0.1:27017
2019-06-26T07:51:02.498+0000 I NETWORK  [js] Successfully connected to 127.0.1.1:27017 (1 connections now open to 127.0.1.1:27017 with a 5 second timeout)
2019-06-26T07:51:02.499+0000 I NETWORK  [js] Marking host 127.0.1.1:27017 as failed :: caused by :: Location40659: can't connect to new replica set master [127.0.1.1:27017], err: AuthenticationFailed: Authentication failed.
2019-06-26T07:51:02.500+0000 I NETWORK  [js] Marking host 127.0.1.1:27017 as failed :: caused by :: Location40659: can't connect to new replica set master [127.0.1.1:27017], err: AuthenticationFailed: Authentication failed.
2019-06-26T07:51:03.003+0000 I NETWORK  [js] Marking host 127.0.1.1:27017 as failed :: caused by :: Location40659: can't connect to new replica set master [127.0.1.1:27017], err: AuthenticationFailed: Authentication failed.
2019-06-26T07:51:03.507+0000 I NETWORK  [js] Marking host 127.0.1.1:27017 as failed :: caused by :: Location40659: can't connect to new replica set master [127.0.1.1:27017], err: AuthenticationFailed: Authentication failed.
2019-06-26T07:51:03.507+0000 E QUERY    [js] Error: can't authenticate against replica set node 127.0.1.1:27017 :: caused by :: can't connect to new replica set master [127.0.1.1:27017], err: AuthenticationFailed: Authentication failed. :
connect@src/mongo/shell/mongo.js:344:17
@(connect):2:6
exception: connect failed
```

Please write some kind of migration guide for single node replicaset without authentication.

**Comment by marcosbc** [comment]
Hi @misuzu, you should also specify the `authSource=admin` as a parameter to the MongoDB connection string, e.g.:

```
$ mongo mongodb://127.0.0.1/default?replicaSet=test123&authSource=admin
```

We'll investigate why this change is needed.

**Comment by MarkPolivchuk** [comment]
Any updates on this? I've run into the same issue

**Comment by alemorcuq** [comment]
Hi, @MarkPolivchuk.

No updates yet, it's in our backlog. We'll notify you when it's fixed. Thank you for your patience.

Regards,
Alejandro

**Comment by javieryanez** [comment]
In the file libmongodb.sh i see this:

```
elif [[ "$MONGODB_REPLICA_SET_MODE" = "primary" ]]; then
           .
           .
           .
            if [[ -z "$MONGODB_ROOT_PASSWORD" ]]; then
                error_message="MONGODB_ROOT_PASSWORD have to be set on a 'primary' node!"
                print_validation_error "$error_message"
            fi
```
If set MONGODB_ROOT_PASSWORD isn't mandatory, in my opinion, this check over MONGODB_ROOT_PASSWORD isn't right and is the cause of this issue.

**Comment by marcosbc** [comment]
You're right @javieryanez. After trying again with the latest MongoDB it is not necessary to use the `authSource` anymore, so your suggestion should be enough. We'll apply it shortly, meaning a fix should be available soon.

It should be enough to connect with the command below:

    $ mongo 'mongodb://localhost/?replicaSet=test123'

**Comment by marcosbc** [comment]
Hi all, the suggestion above has been applied to all versions of the Bitnami MongoDB container. I understand nothing is pending for this issue, so I'll close it. Feel free to add any comment if you continue to find issues.

**Comment by MihaiAnei** [comment]
This is still seems to causes an issue here when trying to deploy a replica set without password:

https://github.com/helm/charts/issues/16450

stable/mongodb chart version 7.8.1. Secondaries start without issues but the master and arbiter don't:
```
/libmongodb.sh: line 780: 4: password is required

 MONGODB_ROOT_PASSWORD have to be set on a 'primary' node! 
```

Any ideas? Should the chart not allow passwordless replicasets?

**Comment by marcosbc** [comment]
You're right @MihaiAnei, it seems we re-added this logic when creating the Bitnami MongoDB Sharded docker image one week after removing, due to a merge conflict. I'll proceeding to remove it again.

**Comment by marcosbc** [comment]
@MihaiAnei A fix was promoted and released in `4.2.3-debian-10-r9`. Unfortunately, due to there not being MongoDB 3.6 and 4.0 binaries available for Debian 10, we are currently unable to release those versions yet. We are internally looking into changing that, probably by going back to Debian 9. If you need one of those versions, you can check out the Oracle Linux images tags `3.6.17-ol-7-r8` or `4.0.16-ol-7-r3`.

You will see that the environment variable `ALLOW_EMPTY_PASSWORD` is also needed, which we decided after checking the existing logic for the other `MONGODB_PASSWORD` environment variable: https://github.com/bitnami/bitnami-docker-mongodb/blob/master/4.2/debian-10/rootfs/libmongodb.sh#L165

It is possible the chart does not support that value, in which case I would report back to the original issue in helm/charts.
